text/0000-conservative-impl-trait.md

```diff
@@ -0,0 +1,553 @@
+  - The type would not be known to implement any other trait, with
+    the exception of OIBITS (aka "auto traits") and default traits like `Sized`.
+  - The type would not be considered equal to the actual underlying type.
+  - The type would not be allowed to appear as the Self type for an `impl` block.
+
+- Because OIBITS like `Send` and `Sync` will leak through an abstract return
+  type, there will be some additional complexity in the compiler due to some
+  non-local type checking becoming necessary.
+
+- The return type has an identity based on all generic parameters the
+  function body is parametrized by, and by the location of the function
+  in the module system. This means type equality behaves like this:
+
+  ```rust
+  fn foo<T: Trait>(t: T) -> @Trait {
+      t
+  }
+
+  fn bar() -> @Trait {
+      123
+  }
+
+  fn equal_type<T>(a: T, b: T) {}
+
+  equal_type(bar(), bar());                      // OK
+  equal_type(foo::<i32>(0), foo::<i32>(0));      // OK
+  equal_type(bar(), foo::<i32>(0));              // ERROR, `@Trait {bar}` is not the same type as `@Trait {foo<i32>}`
+  equal_type(foo::<bool>(false), foo::<i32>(0)); // ERROR, `@Trait {foo<bool>}` is not the same type as `@Trait {foo<i32>}`
+  ```
+
+- The code generation passes of the compiler would not draw a distinction
+  between the abstract return type and the underlying type, just like they don't
+  for generic paramters. This means:
+  - The same trait code would be instantiated, for example, `-> @Any`
+    would return the type id of the underlying type.
+    - Specialization would specialize based on the underlying type.
+
+**Initial limitations**:
+
+- `@Trait` may only be written within the return type of a freestanding or
+  inherent-impl function, not in trait definitions or any non-return type position. They may also not appear
+  in the return type of closure traits or function pointers,
+  unless these are themself part of a legal return type.
+
+  - Eventually, we will want to allow the feature to be used within traits, and
+    like in argument position as well (as an ergonomic improvement over today's generics).
+  - Using `@Trait` multiple times in the same return type would be valid,
+    like for example in `-> (@Foo, @Bar)`.
+
+- The type produced when a function returns `@Trait` would be effectively
+  unnameable, just like closures and function items.
+
+  - We will almost certainly want to lift this limitation in the long run, so
+    that abstract return types can be placed into structs and so on. There are a
+    few ways we could do so, all related to getting at the "output type" of a
+    function given all of its generic arguments.
+
+- The function body cannot see through its own return type, so code like this
+  would be forbidden just like on the outside:
+
+  ```rust
+  fn sum_to(n: u32) -> @Display {
+      if n == 0 {
+          0
+      } else {
+          n + sum_to(n - 1)
+      }
+  }
```

Comment:
Actually I'm not even sure this should type check, it would require `0: u32` being coerced to `@Display` inside the function body, rather than at the return point (so that the whole `if` block has type `@Display`. If we add `return` to at least one branch, I guess it should type check though.